### PR TITLE
[WIP] Continued effort for refactoring and standardizing the Cromwell-tools API

### PR DESCRIPTION
**Pull request by rexwangcc**
I still need to fix a lot of unit tests, but the main part is ready for an early review!

---
Please ensure the following when opening a PR:
- [x] Added or updated tests
- [x] Updated documentation
- [x] Applied style guidelines
- [x] Considered generalizability beyond our own use case

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -32,10 +32,23 @@ Install it like this:
 Usage
 =====
 
+Python API
+----------
+The rest of the package consists of scripts that are meant to be invoked from the command line.
```
Comment (samanehsan) [inline README.rst]: This sentence doesn't make sense to me here since this is the "Python API" section. I think we can remove it!

File: cromwell_tools/_cli.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    wait = subparsers.add_parser(
+            'wait', help='wait help', description='Wait for one or more running workflow to finish.')
+    status = subparsers.add_parser(
+            'status', help='status help', description='Get the status of one or more workflows.')
+    health = subparsers.add_parser(
+            'health', help='health help',
+            description='Check that cromwell is running and that provided authentication is valid')
+
+    # cromwell url and authentication arguments apply to all sub-commands
+    cromwell_sub_commands = [submit, wait, status, health]
```
Comment (samanehsan) [inline cromwell_tools/_cli.py]: The commands available through the CLI are missing "query", "abort" and "metadata". These could be added in a follow-up PR if you prefer since this refactor will already change a bunch of code!

File: cromwell_tools/_cromwell_api.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+        response = cls.requests.post(url=auth.url + cls._workflow_endpoint,
+                                     files=files,
+                                     auth=auth.auth,
+                                     headers=auth.header)
+
+        response.raise_for_status()
+        return response
+
+    @classmethod
+    def wait(cls, workflow_ids, timeout_minutes, auth, poll_interval_seconds=30, verbose=True):
```
Comment (samanehsan) [inline cromwell_tools/_cromwell_api.py]: We can set the default `timeout_minutes` to 120 here so that it is consistent with the behavior of the CLI

File: cromwell_tools/_cli.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    # todo see if this can be moved or if the commands can be populated from above
+    if args['command'] in ('run', 'wait', 'status', 'health'):
+        auth = CromwellAuth.harmonize_credentials(**args)
+        args['auth'] = auth
+    command = getattr(CromwellAPI, args['command'])
+    del args['command']
+    return command, args
+
+
+# this should just getattr from CromwellAPI and call the func with args.
```
Comment (samanehsan) [inline cromwell_tools/_cli.py]: > this should just getattr from CromwellAPI and call the func with args.

It looks like this is already being done at the end of the `parser` function, but it could be simpler to do `getattr(CromwellAPI, args['command'])` in the main function instead of returning it from the parser function.

File: cromwell_tools/_cromwell_api.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+
+        Args:
+            response (requests.Response): A status response object from Cromwell.
+
+        Raises:
+            requests.exceptions.HTTPError: This will be raised when Cromwell returns a response that satisfies
+                400 <= response.status_code < 600.
+        """
+        if not response.ok:
+            raise requests.exceptions.HTTPError('Error Code {0}: {1}'.format(response.status_code, response.text))
```
Comment (samanehsan) [inline cromwell_tools/_cromwell_api.py]: Can you remind me why `_check_and_raise_status` is now configurable? When would users want to use one option over another?

File: cromwell_tools/_cromwell_api.py [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+                else:
+                    query_params.append(
+                            {k: json.dumps(v)} if not isinstance(v, str) else {k: v}
+                    )
+            else:
+                logger.info('{} is not an allowed query key in Cromwell, will be ignored in this query.'.format(k))
+        return query_params
+
+    @staticmethod
+    def _parse_status(response):
```
Comment (samanehsan) [inline cromwell_tools/_cromwell_api.py]: How about renaming this to `_parse_workflow_status`? At first I thought it referred to the HTTP status code

File: cromwell_tools/_cromwell_api.py [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+                return
+
+            time.sleep(poll_interval_seconds)
+
+    @classmethod
+    def release_hold(cls, uuid, auth, raise_for_status=False):
+        """Request Cromwell to release the hold on a workflow.
+
+        It will switch the status of a workflow from ‘On Hold’ to ‘Submitted’ so it can be picked for running. For
+        a workflow that was not submitted with `workflowOnHold = true`, Cromwell will ignore the request.
```
Comment (samanehsan) [inline cromwell_tools/_cromwell_api.py]: > "For a workflow that was not submitted with `workflowOnHold = true`, Cromwell will ignore the request."

Is this still the intended behavior for this function? Cromwell will throw a 403 error if the workflow is not on hold, so we would have to handle that error here.

File: cromwell_tools/_cromwell_api.py [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+
+    @classmethod
+    def query(cls, query_dict, auth, raise_for_status=False):
+        """
+        TODO: Given that Cromwell-as-a-Service blocks a set of features that are available in Cromwell, e.g. 'labelor',
+        for security concerns, the first iteration of this API doesn't come up with the advanced query keys of the
+        Cromwell except a set of necessary ones. However, we need to implement this for completeness and keep an eye
+        on the compatibility between CaaS and Cromwell.
+
+        _All of the query keys will be used in an OR manner, except the keys within `labels`, which are defined in
```
Comment (samanehsan) [inline cromwell_tools/_cromwell_api.py]: Why are there underscores around this sentence?

**Comment by samanehsan** [comment]
Is this PR up to date with the master branch? I noticed it is missing this function to validate workflows: https://github.com/broadinstitute/cromwell-tools/blob/master/cromwell_tools/cromwell_tools.py#L438

File: cromwell_tools/_cromwell_api.py [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+    pass
+
+
+# TODO: use functools partial for get, post (set the authenticate commands)
+class CromwellAPI(object):
+    """Contains a set of classmethods that implement interfaces to cromwell REST API endpoints.
+    """
+
+    # TODO: remove this if it is not critical for mocking
+    requests = requests
```
Comment (samanehsan) [inline cromwell_tools/_cromwell_api.py]: Why is this necessary?

File: cromwell_tools/tests/test_cromwell_api.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+auth_types = setup_auth_types()
+
+
+@pytest.fixture(scope='module', params=auth_types.values(), ids=auth_types.keys())
+def auth_types(request):
+    return request.param
+
+
+def test_cromwell_auth(auth_types):
```
Comment (samanehsan) [inline cromwell_tools/tests/test_cromwell_api.py]: This test is duplicated in `test_cromwell_auth.py`, which also includes similar set-up code so I think we can remove this file.

**Comment by samanehsan** [comment]
Done with a first-pass review @rexwangcc! :octocat:

File: cromwell_tools/_cromwell_api.py (hunk 3 above)
Comment (rexwangcc) [reply cromwell_tools/_cromwell_api.py]: 👍 Will do!

File: cromwell_tools/_cromwell_api.py (hunk 7 above)
Comment (rexwangcc) [reply cromwell_tools/_cromwell_api.py]: Great catch, this docstring is lying, the post request will raise 403 and, this function will return the errored response or raise the error to the users(if raise_for_status==True).

File: cromwell_tools/_cromwell_api.py (hunk 8 above)
Comment (rexwangcc) [reply cromwell_tools/_cromwell_api.py]: 🤔 IDK... Let me remove them!

File: cromwell_tools/_cromwell_api.py (hunk 6 above)
Comment (rexwangcc) [reply cromwell_tools/_cromwell_api.py]: 👍

File: cromwell_tools/_cromwell_api.py (hunk 5 above)
Comment (rexwangcc) [reply cromwell_tools/_cromwell_api.py]: So in general, there are 2 use cases:
- Users(such as Lira) wants to have their own retry policy based on the response code, and they might also want to deal with the response body directly. In these cases, this library should give them the raw/transparent response from Cromwell. (_check_and_raise_status=False by default)
- There are other situations that users(such as who invoke this function from CLI) prefer to know more about the error message than dealing with the errors, and they don't care about who raises the error(Cromwell or cromwell-tools), as long as the error messages are friendly and reasonable. In these cases, we will need cromwell-tools and catch and raise the errors in a proper way. (_check_and_raise_status=True)

File: cromwell_tools/_cromwell_auth.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        else:
+            raise ValueError("url must be an str that points to an http(s) endpoint.")
+
+    @staticmethod
+    def _validate_connection(header, auth, url):
+        warnings.warn("This function doesn't work for Cromwell until Cromwell adds the endpoint for validating users'"
+                      "Auth information, e.g. a Who am I endpoint.", DeprecationWarning)
+        return None
+
+        dummy_auth = namedtuple('dummy_auth', ['header', 'auth', 'url'])
```
[the remark this answers is not in the document]
Comment (rexwangcc) [reply cromwell_tools/_cromwell_auth.py]: Good point, I'm taking it out for now! It was there in the stale PR but I realized that it can't work, since the `/health` endpoint of Cromwell it's using now doesn't require auth at all. Returning `None` is like a short-out workaround, but I agree that taking it out and make a ticket is a better way!!

File: README.rst [hunk 12]
```diff
@@ -32,10 +32,21 @@ Install it like this:
 Usage
 =====
 
+Python API
+----------
+In Python, you can then import the package with:
```
Comment (samanehsan) [inline README.rst]: I think this would make more sense as `In Python, you can import the package with:`

**Comment by rexwangcc** [comment]
Closing with https://github.com/broadinstitute/cromwell-tools/pull/43
